module04/03-Panda-Social-Network/solution.py

```diff
@@ -0,0 +1,151 @@
+import re
+
+
+class Panda:
+    """Panda which is user of :class:`PandaSocialNetwork` and has basic attributes as name,
+    email, gender and list of friends"""
+
+    def __init__(self, name, email, gender):
+        self.validate_input(name, email, gender)
+        self.panda_name = name
+        self.panda_email = email
+        self.panda_gender = gender
+        self.friend_list = []
+
+    def __str__(self):
+        return "%s panda named %s with email: %s" % (self.panda_gender.capitalize(), self.panda_name, self.panda_email)
+
+    def __eq__(self, other):
+        if isinstance(other, Panda):
+            return self.__key() == other.__key()
+        raise NotImplemented
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __key(self):
+        return self.panda_name, self.panda_email, self.panda_gender
+
+    def validate_input(self, name, email, gender):
+        if not name or not email or not gender:
+            raise ValueError("Input not provided")
+        if not self.is_valid_email(email):
+            raise ValueError("Not valid email provided")
+
+    @staticmethod
+    def is_valid_email(email):
```

Comment:
like above, this should be a private method. and having it static does not bring any immediate value, because you are calling it just once (on line 32), and `self` there is an instance